engineapi/proxy.go

```diff
@@ -0,0 +1,291 @@
+package engineapi
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	imclient "github.com/longhorn/longhorn-instance-manager/pkg/client"
+	imutil "github.com/longhorn/longhorn-instance-manager/pkg/util"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+const (
+	ProxyPollInterval = 5 * time.Second
+)
+
+type EngineClientProxy interface {
+	EngineClient
+
+	IsGRPC() bool
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Stop(string) error
+	Ping() error
+}
+
+// TODO: replace with EngineClientProxy interface later
+type Client interface {
+	IsGRPC() bool
+	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
+	Stop(string) error
+	Ping() error
+}
+
+type ProxyHandler struct {
+	Handle map[string]Client
+
+	logger logrus.FieldLogger
+
+	ds *datastore.DataStore
+
+	lock *sync.RWMutex
+}
+
+func NewProxyHandler(logger logrus.FieldLogger, ds *datastore.DataStore) *ProxyHandler {
+	return &ProxyHandler{
+		Handle: make(map[string]Client),
+		logger: logger,
+		ds:     ds,
+		lock:   &sync.RWMutex{},
+	}
+}
+
+func getLoggerForProxy(logger logrus.FieldLogger, image string) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"instanceManagerImage": image,
+		},
+	)
+}
+
+func (h *ProxyHandler) GetClient(im *longhorn.InstanceManager, fallBack interface{}) (c Client, err error) {
+	image := ""
+	if im == nil {
+		image, err = h.ds.GetSettingValueExisted(types.SettingNameDefaultInstanceManagerImage)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		image = im.Spec.Image
+	}
+	log := getLoggerForProxy(h.logger, image)
+
+	shouldFallBack := false
+
+	if im != nil {
+		if err := CheckInstanceManagerProxyCompatibility(im); err != nil {
+			log.WithError(err).Warnf("Use fallback client")
+			shouldFallBack = true
+		}
+	}
+
+	if shouldFallBack {
+		if fallBack == nil {
+			return nil, errors.Errorf("missing fallback client for proxy")
+		}
+
+		if obj, ok := fallBack.(*EngineBinary); ok {
+			return obj, nil
+		}
+
+		if obj, ok := fallBack.(*BackupTargetClient); ok {
+			return obj, nil
+		}
+
+		return nil, errors.Errorf("BUG: invalid fallback client for proxy: %v", fallBack)
+	}
+
+	client, exist := h.Handle[image]
+	if exist {
+		err := client.Ping()
+		if err == nil {
+			return client, nil
+		}
+	}
+
+	return h.NewClient(log, image, im)
+}
+
+func (h *ProxyHandler) NewClient(log logrus.FieldLogger, image string, im *longhorn.InstanceManager) (Client, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	existClient, exist := h.Handle[image]
+	if exist {
+		err := existClient.Ping()
```

Comment:
What's timeout for the ping? is it possible to cause some latency for `NewClient` due to timeout ping? or it's acceptable?

Reply:
ATM the client has 3 minutes context timeout. I will create another issue for enhancement.

https://github.com/longhorn/longhorn-manager/pull/1212#discussion_r869998036